File: clientPython/stats.py
import math
import time


class RttStats:
    def __init__(self) -> None:
        self._rtts: list[float] = []

    def add(self, rtt_ms: float) -> None:
        self._rtts.append(rtt_ms)

    @property
    def received(self) -> int:
        return len(self._rtts)

    def print_summary(self, sent: int, start_time: float, host: str) -> None:
        elapsed_ms = (time.perf_counter() - start_time) * 1000
        loss = 100.0 * (1.0 - self.received / sent) if sent else 0.0
        print(f"\n--- {host} ping statistics ---")
        print(
            f"{sent} packets transmitted, {self.received} received, "
            f"{loss:.1f}% packet loss, time {elapsed_ms:.0f}ms"
        )
        if self.received > 0:
            min_rtt = min(self._rtts)
            avg_rtt = sum(self._rtts) / self.received
            max_rtt = max(self._rtts)
            mdev = math.sqrt(sum((x - avg_rtt) ** 2 for x in self._rtts) / self.received)
            print(
                f"rtt min/avg/max/mdev = "
                f"{min_rtt:.3f}/{avg_rtt:.3f}/{max_rtt:.3f}/{mdev:.3f} ms"
            )

File: clientRust/src/stats.rs
use std::time::Duration;

pub struct RttStats {
    rtts: Vec<f64>,
}

impl RttStats {
    pub fn new() -> Self {
        Self { rtts: Vec::new() }
    }

    pub fn add(&mut self, rtt_ms: f64) {
        self.rtts.push(rtt_ms);
    }

    pub fn received(&self) -> usize {
        self.rtts.len()
    }

    pub fn print_summary(&self, sent: u32, elapsed: Duration, host: &str) {
        let received = self.received();
        let loss = if sent > 0 {
            100.0 * (1.0 - received as f64 / sent as f64)
        } else {
            0.0
        };
        println!("\n--- {host} ping statistics ---");
        println!(
            "{sent} packets transmitted, {received} received, \
             {loss:.1}% packet loss, time {}ms",
            elapsed.as_millis()
        );
        if received > 0 {
            let min = self.rtts.iter().cloned().fold(f64::INFINITY, f64::min);
            let max = self.rtts.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
            let avg = self.rtts.iter().sum::<f64>() / received as f64;
            let mdev = (self.rtts.iter().map(|&x| (x - avg).powi(2)).sum::<f64>()
                / received as f64)
                .sqrt();
            println!("rtt min/avg/max/mdev = {min:.3}/{avg:.3}/{max:.3}/{mdev:.3} ms");
        }
    }
}
